functional_programming_tests/test/with_dependency_injection.cpp

```diff
@@ -0,0 +1,528 @@
+   * @param data The 2D vector to construct the map from
+   */
+  Map(std::vector<std::vector<T>> data) : data_{data} {};
+  T at(Position const& pos) const { return data_.at(pos.y).at(pos.x); }
+  std::vector<std::vector<T>>& get_data() { return data_; }
+
+ private:
+  std::vector<std::vector<T>> data_;
+};
+
+struct PathGenerator {
+  struct MiddlewareHandle {
+    // Define map service callback type
+    using SetMapServiceCallback = std::function<void(
+        std::shared_ptr<example_srvs::srv::SetMap::Request> const request,
+        std::shared_ptr<example_srvs::srv::SetMap::Response> response)>;
+
+    // Define path generation service callback type
+    using GeneratePathServiceCallback = std::function<void(
+        std::shared_ptr<example_srvs::srv::GetPath::Request> const request,
+        std::shared_ptr<example_srvs::srv::GetPath::Response> response)>;
+    /**
+     * @brief Ensure public dtor is virtual
+     * @note
+     * http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-virtual
+     */
+    virtual ~MiddlewareHandle() = default;
+
+    /**
+     * @brief Register a callback for the set map service
+     * @param callback The callback to register
+     */
+    virtual void register_set_map_service(SetMapServiceCallback callback) = 0;
+
+    /**
+     * @brief Register a callback for the generate path service
+     * @param callback The callback to register
+     */
+    virtual void register_generate_path_service(
+        GeneratePathServiceCallback callback) = 0;
+
+    /**
+     * @brief Log error message
+     * @param msg The message to log
+     */
+    virtual void log_error(std::string const& msg) = 0;
+
+    /**
+     * @brief Log info message
+     * @param msg The message to log
+     */
+    virtual void log_info(std::string const& msg) = 0;
+  };
+
+  /**
+   * @brief Builds paths given a map
+   * @param mw The middleware handle for interacting with services
+   */
+  PathGenerator(std::unique_ptr<MiddlewareHandle> mw) : mw_{std::move(mw)} {
+    // Register the set map service
+    mw_->register_set_map_service(
+        [this](
+            std::shared_ptr<example_srvs::srv::SetMap::Request> const request,
+            std::shared_ptr<example_srvs::srv::SetMap::Response> response) {
+          // Set the map to generate the path from
+          response->success.data = costmap_setter(request->map);
+        });
+    // Register the generate path service
+    mw_->register_generate_path_service(  // Trivial comment to enfore params on
+                                          // new line
+        [this](
+            std::shared_ptr<example_srvs::srv::GetPath::Request> const request,
+            std::shared_ptr<example_srvs::srv::GetPath::Response> response) {
+          if (map_.get_data().size() == 0) {
+            mw_->log_error("MAP IS EMPTY!!");
+            response->success.data = false;
+            response->path = std_msgs::msg::UInt8MultiArray();
+            return;
```

Comment:
This is where you added the return right?

Reply:
yessir